File: leetcode/binary-tree-right-side-view.py
from typing import List, Optional

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if root is None:
            return []
        else:
            result = []
            elem = (root, 1)
            queue = [elem]
            while len(queue) != 0:
                cur = queue.pop(0)
                print('cur: ', cur[0].val)
                if cur[1] != elem[1]:
                    print('boom: ', elem[0].val)
                    result.append(elem[0].val)
                if cur[0].left != None:
                    queue.append((cur[0].left, cur[1] + 1))
                if cur[0].right != None:
                    queue.append((cur[0].right, cur[1] + 1))
                elem = cur
            result.append(elem[0].val)
            return result

    def rightSideView_wrong(self, root: Optional[TreeNode]) -> List[int]:
        if root is None:
            return []
        else:
            res = [root.val]
            next = self.rightSideView(root.right)
            return res + next


# === Test Case 1 ===#
t1 = TreeNode(1)
t2 = TreeNode(2)
t3 = TreeNode(3)
t4 = TreeNode(5)
t5 = TreeNode(4)
t1.left = t2
t1.right = t3
t2.right = t4
t3.right = t5

# === Test Case 2 ===#
# t1 = TreeNode(1)
# t2 = TreeNode(3)
# t1.right = t2

# === Test Case 3 ===#
# t1 = None

# === Test Case 4 ===#
# t1 = TreeNode(1)
# t2 = TreeNode(2)
# t1.left = t2

sol = Solution().rightSideView(t1)

print(sol)

File: leetcode/BinaryTreeRightSideView.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class BinaryTreeRightSideView {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }


    class Solution {
        class DepthTreeNode {
            TreeNode t;
            int d;

            public DepthTreeNode(TreeNode t, int d) {
                this.t = t;
                this.d = d;
            }
        }

        public List<Integer> rightSideView(TreeNode root) {
            List<Integer> result = new ArrayList<>();
            if (root == null) {
                return result;
            }

            Queue<DepthTreeNode> queue = new LinkedList<>();
            queue.add(new DepthTreeNode(root, 1));
            int prevH = 1;
            int prevV = root.val;
            while (!queue.isEmpty()) {
                DepthTreeNode current = queue.remove();
                if (prevH != current.d) {
                    result.add(prevV);
                }
                if (current.t.left != null) {
                    queue.add(new DepthTreeNode(current.t.left, current.d + 1));
                }
                if (current.t.right != null) {
                    queue.add(new DepthTreeNode(current.t.right, current.d + 1));
                }
                prevH = current.d;
                prevV = current.t.val;
            }
            result.add(prevV);

            return result;
        }

        public List<Integer> rightSideView2(TreeNode root) {
            List<Integer> outResult = new ArrayList<>();
            if (root == null) {
                return outResult;
            }
            nextLevel(root, outResult, 0);
            return outResult;
        }


        public void nextLevel(TreeNode node, List<Integer> outResult, int level) {
            if (node == null) {
                return;
            }
            if (outResult.size() == level) {
                outResult.add(node.val);
            }
            nextLevel(node.right, outResult, level + 1);
            nextLevel(node.left, outResult, level + 1);
        }

    }
}
